File: my-folder/0871-keys-and-rooms/solution.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        visited[0] = true;
        dfs(rooms, 0, visited);
        for (boolean visit : visited) {
            if (visit == false) return false;
        }
        return true;
    }

    private void dfs(List<List<Integer>> rooms, int room, boolean[] visited) {
        for (int key : rooms.get(room)) {
            if (!visited[key]) {
                visited[key] = true;
                dfs(rooms, key, visited);
            }
        }
    }
}

File: my-folder/0871-keys-and-rooms/solution.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False] * len(rooms)
        visited[0] = True
        def dfs(rooms, current, visited):
            for key in rooms[current]:
                if not visited[key]:
                    visited[key] = True
                    dfs(rooms, key, visited)
        
        dfs(rooms, 0, visited)

        for visit in visited:
            if not visit: return False
        
        return True
